repository_service_tuf_worker/repository.py

```diff
@@ -471,39 +557,33 @@ def remove_targets(
         deleted_targets: List[str] = []
         not_found_targets: List[str] = []
 
-        # Group target files by responsible 'bins' roles.
-        bin_target_groups: Dict[str, List[str]] = {}
+        # Group target files by responsible 'bins' delegated roles.
+        # This will be used to by `publish_targets`
+        bin_targets: Dict[str, List[targets_models.RSTUFTargets]] = {}
         for target in targets:
             bins_name = self._get_path_succinct_role(target)
-            if bins_name not in bin_target_groups:
-                bin_target_groups[bins_name] = []
+            db_target = targets_crud.read_by_path(self._db, target)
+            if db_target is None or (
+                db_target.action == targets_schema.TargetAction.REMOVE
+                and db_target.published is True
+            ):
+                # not found targets or targets already remove action and
+                # published are not found.
+                not_found_targets.append(target)
+            else:
+                db_target = targets_crud.update_action_remove(
+                    self._db, db_target
+                )
+                deleted_targets.append(target)
 
-            bin_target_groups[bins_name].append(target)
+                if bins_name not in bin_targets:
+                    bin_targets[bins_name] = []
 
-        # Update target file info in responsible 'bins' roles, bump
-        # version and expiry and sign and persist.
-        targets_meta = []
-        with self._redis.lock("TUF_BINS_HASHED"):
-            for bins_name, paths in bin_target_groups.items():
-                bins_role = self._load(bins_name)
-                for path in paths:
-                    if path in bins_role.signed.targets:
-                        bins_role.signed.targets.pop(path)
-                        deleted_targets.append(path)
-                        self._bump_expiry(bins_role, BINS)
-                        self._bump_version(bins_role)
-                        self._sign(bins_role, BINS)
-                        self._persist(bins_role, bins_name)
-                        targets_meta.append(
-                            (bins_name, bins_role.signed.version)
-                        )
-
-                    else:
-                        not_found_targets.append(path)
+                bin_targets[bins_name].append(db_target)
 
-        if len(targets_meta) > 0:
-            self._add_to_unpublished_metas(targets_meta)
-            self._publish_meta_state(targets_meta, update_state)
+        if len(deleted_targets) > 0:
```

Comment:
I noticed you are keeping track of the targets for deletion in `deleted_targets` in `remove_targets`, but you don't keep track of the targets for addition in add_targets.
Why is that and what's the difference between `add_targets` and `remove_targets`?

Reply:
The difference is that ``add_targets`` does not constrain whether the target exists. Once the target data in the payload is valid, it is ok.

Removing targets, there is a constraint if the target exists or not.
A user can require to ``remove_targets`` and part to all does not exist.

In the task response, we tell the user the deleted/removed targets and the not found targets.

Reply:
Ahaa, yes, I see what you mean.